### ISSUE TITLE
An ability to chain various plugins alongside with OVN CNI

### ISSUE THREAD

**Issue by girishmg** [ovn-org/ovn-kubernetes#786]
We need to provide a way to chain plugins such as portMapping, bandwidth, tuning, and so on along with OVN CNI. This is done through network configuration lists and it provides a mechanism to run multiple CNI plugins for a single container in a defined order.

Currently, we support setting ingress/egress bandwidth for the Pod’s interface by directly extracting that information from the POD spec and implementing the QoS in the OVN CNI itself.
We don’t support k8s host port functionality at all.

Ideally, we should support this by having our CNI configuration to be a list like below.

```
{
  "name": "k8s-pod-network",
  "cniVersion": "0.3.1",
  "plugins": [
    {
      "type": "ovn-k8s-cni-overlay",
      "ipam": {},
      "dns": {}
    },
    {
      "type": "portmap",
      "snat": true,
      "capabilities": {"portMappings": true }
    },
    {
     “type”: “ovn_bandwidth”,
     "capabilities": {"bandwidth": true}
    },
    {
      “type”: “tunning”,
      "capabilities": {"bandwidth": true}
    }
  ]
}
```

We should capture the OVN K8s CNI configuration in the config map and use that on each of the node. Currently, we use the following json which is hardcoded.

{
    "cniVersion": "0.3.1",
    "dns": {},
    "ipam": {},
    "name": "ovn-kubernetes",
    "type": "ovn-k8s-cni-overlay"
}

Instead, we should get this from the ConfigMap so that one can add/remove to/from the CNI plugins chain

**Comment by girishmg** [comment]
>     },
>     {
>      “type”: “ovn_bandwidth”,
>      "capabilities": {"bandwidth": true}
>     },
>     {

We could move the above capability into the OVN CNI itself like below

```
{
  "name": "k8s-pod-network",
  "cniVersion": "0.3.1",
  "plugins": [
    {
      "type": "ovn-k8s-cni-overlay",
      "ipam": {},
      "dns": {}
     "capabilities": {"bandwidth": true}
    },
```

That way, when `ovn-k8s-cni-overlay` is executed for CMD_ADD it will already have the bandwidth information in the runtimeConfig map. We can use the information in the map to setup the bandwidth instead of pulling the information from the pod spec.

**Comment by dcbw** [comment]
@girishmg don't we handle port mapping, bandwidth, tuning, etc internally inside ovn-kubernetes? eg portmapping is only for hostPorts, which we handle already. bandwidth we already do via QoS too. Tuning is just used to change the MAC address, which we'd get through https://github.com/ovn-org/ovn-kubernetes/pull/825 or similar changes.

**Comment by girishmg** [comment]
@dcbw  I don't think we handle `portmapping` today in ovn-k8s. At least not with `dockershim` and `shared` gateway mode.

Ideally, we should be handling the `bandwidth` outside of the OVN CNI plugin. We will have to write a new CNI plugin called ovn_bandwidth which we will chain with ovn_k8s_cni_overlay. When ovn_bandwidth gets called by the `kubelet`, then the new plugin will read ingress/egress values and apply it on the interface using ovs-* utilities. Similarly for other capabilities out there.

Supporting this interface, which other CNI plugins (eg., calico) do, will give some sort of familiarity to operators who are moving from other CNI plugins to ours (or vice-versa).
